src/pudl_archiver/depositors/zenodo.py

```diff
@@ -0,0 +1,319 @@
+"""Handle all deposition actions within Zenodo."""
+import json
+import logging
+from typing import BinaryIO
+
+import aiohttp
+import semantic_version  # type: ignore
+
+from pudl_archiver.zenodo.entities import Deposition, DepositionMetadata
+
+logger = logging.getLogger(f"catalystcoop.{__name__}")
+
+
+class ZenodoClientException(Exception):
+    """Captures the JSON error information from Zenodo."""
+
+    def __init__(self, kwargs):
+        """Constructor.
+
+        Args:
+            kwargs: dictionary with "response" mapping to the actual
+                aiohttp.ClientResponse and "json" mapping to the JSON content.
+        """
+        self.kwargs = kwargs
+        self.status = kwargs["response"].status
+        self.message = kwargs["json"].get("message", {})
+        self.errors = kwargs["json"].get("errors", {})
+
+    def __str__(self):
+        """The JSON has all we really care about."""
+        return f"ZenodoClientException({self.kwargs['json']})"
+
+    def __repr__(self):
+        """But the kwargs are useful for recreating this object."""
+        return f"ZenodoClientException({repr(self.kwargs)})"
+
+
+class ZenodoDepositor:
+    """Act on depositions & deposition files within Zenodo."""
+
+    def __init__(
+        self,
+        upload_key: str,
+        publish_key: str,
+        session: aiohttp.ClientSession,
+        sandbox: bool = True,
+    ):
+        """Create a new ZenodoDepositor.
+
+        Args:
+            upload_key: the Zenodo API key that gives you upload rights.
+            publish_key: the Zenodo API key that gives you publish rights.
+            session: HTTP handler - we don't use it directly, it's wrapped in self.request.
+            sandbox: whether to hit the sandbox Zenodo instance or the real one. Default True.
+        """
+        self.auth_write = {"Authorization": f"Bearer {upload_key}"}
+        self.auth_actions = {"Authorization": f"Bearer {publish_key}"}
+        self.request = self._make_requester(session)
+
+        self.sandbox = sandbox
+
+        if sandbox:
+            self.api_root = "https://sandbox.zenodo.org/api"
+        else:
+            self.api_root = "https://zenodo.org/api"
+
+    def _make_requester(self, session):
+        """Wraps our session requests with some Zenodo-specific error handling."""
+
+        async def requester(
+            method: str, url: str, log_label: str, parse_json: bool = True, **kwargs
+        ) -> dict | aiohttp.ClientResponse:
+            """Make requests to Zenodo.
+
+            Args:
+                method: HTTP method - "GET", "POST", etc.
```

Comment:
Could be nice to use an `Enum` or `Literal` to constrain this

Reply:
Woah, didn't know about `Literal` ! Nice.